app/models/custom_field/calculated_value.rb

```diff
@@ -73,29 +75,75 @@ def formula_string
       formula ? formula.fetch("formula", "") : ""
     end
 
+    def usable_custom_field_references_for_formula
+      scope = type == "ProjectCustomField" ? ProjectCustomField : CustomField
+
+      scope
+        .where(field_format: FIELD_FORMATS_FOR_FORMULA)
+        # Disallow the current custom field to avoid circular references
+        .where.not(id: id)
+        .visible
+    end
+
+    def validate_referenced_custom_fields_allowed
+      # We can only validate used custom fields from a high-level perspective, since at this point in validation,
+      # we do not have a project context to check against. So we cannot check if the custom fields are actually
+      # enabled for a project and visible to a non-admin user.
+      formula_cfs = CustomField.where(id: cf_ids_used_in_formula(formula_string)).pluck(:id)
```

Comment:
I think it is better to just use result of `cf_ids_used_in_formula(formula_string)`, as ids of non existing custom fields are also wrong
```suggestion
      formula_cfs = cf_ids_used_in_formula(formula_string)
```